ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java

```diff
@@ -208,12 +208,8 @@ private PullQueryRow pollRow() {
 
     polled.handler.handle(new SucceededFuture<>(null, null));
 
-    if (monitor.enterIf(atHalfCapacity)) {
-      try {
-        drainHandler.forEach(h -> h.handle(null));
-      } finally {
-        monitor.leave();
-      }
+    if (isDone() || size() <= queueCapacity / 2) {
```

Comment:
Why did you remove the monitor guard?

Reply:
Firstly, it's not clear that we need it, because the drain handler doesn't modify the state of the WriteStream - it only modifies the state of the RecordParser, and the RecordParser now has a separate monitor.

More importantly, the monitor now can cause a deadlock:
 - `RecordParser.handle` acquires the new monitor that I introduced, and subsequently calls `PullQueryWriteStream.handle`, which attempts to acquire the `monitor` in this context.
 - `PullQueryWriteStream.pollRow` acquires `monitor` and calls `RecordParser.resume` which attempts to acquire the monitor that I introduced.

So we remove aquiring `monitor` here.

Reply:
@cadonna and I tried running it on the context of the Read Stream, but failed to fix it this way. The problem was that the pull queries would not make progress, and I couldn't determine what was causing it. Today I gave this another look, and I was able to figure it our -- we are calling the drain handler wayyyy to often, so we are clogging the netty event loop completely with drain handler calls. The pull queries make progress, but the netty event loops are essentially at 100% CPU just going through drain handler calls. I think the correct fix is to call the drain handler only exactly once - the pipe implementation will re-register its drain handler when the `WriteStream` reaches capacity. @AlanConfluent could you have a look at the alternative fix here: https://github.com/confluentinc/ksql/pull/10077